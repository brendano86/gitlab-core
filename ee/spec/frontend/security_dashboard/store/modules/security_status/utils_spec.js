import {
  getMostSevereVulnerabilityType,
  getProjectData,
  getSeverityGroupForType,
  getSeverityGroups,
  getVulnerabilityCount,
  groupBySeverity,
} from 'ee/security_dashboard/store/modules/security_status/utils';

const createMockProjectWithZeroVulnerabilities = () => ({
  id: 'id',
  full_name: 'full_name',
  full_path: 'full_path',
  critical_vulnerability_count: 0,
  high_vulnerability_count: 0,
  medium_vulnerability_count: 0,
  low_vulnerability_count: 0,
  unknown_vulnerability_count: 0,
});

const createMockProjectWithVulnerabilities = (count = 1) => (...vulnerabilityLevels) => ({
  ...createMockProjectWithZeroVulnerabilities(),
  ...(vulnerabilityLevels
    ? vulnerabilityLevels.reduce(
        (levels, level) => ({
          ...levels,
          [`${level}_vulnerability_count`]: count,
        }),
        {},
      )
    : {}),
});

const createMockProjectWithOneVulnerability = createMockProjectWithVulnerabilities(1);

describe('SeverityLevels store utils', () => {
  describe('getMostSevereVulnerabilityType', () => {
    it.each`
      vulnerabilityTypesInProject                         | expectedType
      ${['critical', 'high', 'unknown', 'medium', 'low']} | ${'critical'}
      ${['high', 'unknown', 'medium', 'low']}             | ${'high'}
      ${['unknown', 'medium', 'low']}                     | ${'unknown'}
      ${['medium', 'low']}                                | ${'medium'}
      ${['low']}                                          | ${'low'}
      ${[]}                                               | ${'none'}
    `(
      "given $vulnerabilityTypesInProject returns '$expectedType'",
      ({ vulnerabilityTypesInProject, expectedType }) => {
        const mockProject = createMockProjectWithOneVulnerability(...vulnerabilityTypesInProject);

        expect(getMostSevereVulnerabilityType(mockProject)).toEqual({
          displayName: expectedType,
          name: expectedType,
        });
      },
    );
  });

  describe('getProjectData', () => {
    it('takes a project and its most severe vulnerability type and returns the data used to render the project', () => {
      expect(
        getProjectData(createMockProjectWithZeroVulnerabilities(), {
          displayName: 'foo',
          name: 'bar',
        }),
      ).toMatchSnapshot();
    });
  });

  describe('getSeverityGroupForType', () => {
    it.each`
      vulnerabilityType | expectedSeverityGroup
      ${'critical'}     | ${'F'}
      ${'high'}         | ${'E'}
      ${'unknown'}      | ${'E'}
      ${'medium'}       | ${'D'}
      ${'low'}          | ${'B'}
      ${'none'}         | ${'A'}
    `(
      'returns $expectedSeverityGroup for the given $vulnerabilityType',
      ({ vulnerabilityType, expectedSeverityGroup }) => {
        expect(getSeverityGroupForType(vulnerabilityType).name).toBe(expectedSeverityGroup);
      },
    );
  });

  describe('getSeverityGroups', () => {
    it("returns a map with the given severity levels containing an empty 'projects' array", () => {
      const severityLevels = [
        {
          name: 'fooName',
          displayName: 'fooDisplayName',
          description: 'fooDescription',
        },
        {
          name: 'barName',
          displayName: 'barDisplayName',
          description: 'barDescription',
        },
      ];

      const groups = getSeverityGroups(severityLevels);

      expect(groups.get('fooName')).toStrictEqual({
        name: 'fooDisplayName',
        description: 'fooDescription',
        projects: [],
      });

      expect(groups.get('barName')).toStrictEqual({
        name: 'barDisplayName',
        description: 'barDescription',
        projects: [],
      });
    });
  });

  describe('getVulnerabilityCount', () => {
    it.each`
      vulnerabilityType | vulnerabilityCount
      ${'critical'}     | ${1}
      ${'high'}         | ${2}
      ${'medium'}       | ${3}
      ${'low'}          | ${4}
      ${'unknown'}      | ${5}
    `(
      "returns the correct count for '$vulnerabilityType' vulnerabilities",
      ({ vulnerabilityType, vulnerabilityCount }) => {
        const project = createMockProjectWithVulnerabilities(vulnerabilityCount)(vulnerabilityType);

        expect(getVulnerabilityCount(vulnerabilityType, project)).toBe(vulnerabilityCount);
      },
    );
  });

  describe('groupBySeverity', () => {
    it('takes an array of projects containing vulnerability data and groups them by severity level', () => {
      const projectsWithVulnerabilities = [
        createMockProjectWithOneVulnerability('critical'),
        createMockProjectWithOneVulnerability('high'),
        createMockProjectWithOneVulnerability('unknown'),
        createMockProjectWithOneVulnerability('medium'),
        createMockProjectWithOneVulnerability('low'),
        createMockProjectWithZeroVulnerabilities(),
      ];

      const projectsGroupedBySeverityLevel = groupBySeverity(
        Object.values(projectsWithVulnerabilities),
      );

      expect(projectsGroupedBySeverityLevel).toMatchSnapshot();
    });
  });
});
