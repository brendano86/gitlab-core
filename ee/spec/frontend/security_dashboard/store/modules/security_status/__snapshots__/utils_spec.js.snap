// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeverityLevels store utils getProjectData takes a project and its most severe vulnerability type and returns the data used to render the project 1`] = `
Object {
  "id": "id",
  "mostSevere": Object {
    "count": undefined,
    "name": "foo",
  },
  "name": "full_name",
  "path": "full_path",
}
`;

exports[`SeverityLevels store utils groupBySeverity takes an array of projects containing vulnerability data and groups them by severity level 1`] = `
Array [
  Object {
    "description": "Some description for F",
    "name": "F",
    "projects": Array [
      Object {
        "id": "id",
        "mostSevere": Object {
          "count": 1,
          "name": "critical",
        },
        "name": "full_name",
        "path": "full_path",
      },
    ],
  },
  Object {
    "description": "Some description for E",
    "name": "E",
    "projects": Array [
      Object {
        "id": "id",
        "mostSevere": Object {
          "count": 1,
          "name": "high",
        },
        "name": "full_name",
        "path": "full_path",
      },
      Object {
        "id": "id",
        "mostSevere": Object {
          "count": 1,
          "name": "unknown",
        },
        "name": "full_name",
        "path": "full_path",
      },
    ],
  },
  Object {
    "description": "Some description for D",
    "name": "D",
    "projects": Array [
      Object {
        "id": "id",
        "mostSevere": Object {
          "count": 1,
          "name": "medium",
        },
        "name": "full_name",
        "path": "full_path",
      },
    ],
  },
  Object {
    "description": "Some description for B",
    "name": "B",
    "projects": Array [
      Object {
        "id": "id",
        "mostSevere": Object {
          "count": 1,
          "name": "low",
        },
        "name": "full_name",
        "path": "full_path",
      },
    ],
  },
  Object {
    "description": "Some description for A",
    "name": "A",
    "projects": Array [
      Object {
        "id": "id",
        "mostSevere": Object {
          "count": undefined,
          "name": "none",
        },
        "name": "full_name",
        "path": "full_path",
      },
    ],
  },
]
`;
